### PR TITLE
support nested_tensor * scalar

### PR DESCRIPTION
**Pull request by YifanShenSZ**
In transformer, the scale step in attention has a `nested_tensor / scalar` operation. There are two ways to support that:
1. directly support `nested_tensor / scalar`:
* pro: straightforward, good UX
* con: is dispatching `mul(nested tensor, regular tensor)` a good practice?
2. let user manually convert `scalar` to `nested_scalar = torch.nested_tensor([broadcast_scalar])`
* pro: dispatcher only has to deal with `mul(nested tensor, nested tensor)`
* con: confusing manual conversions, bad UX

### CODE REVIEW

**Comment by bdhirsh** [comment]
Stuff like op(cuda_tensor, scalar_tensor) is allowed, so I think mixing scalar tensors with nested tensors should be allowed too. FWIW, that should work fine with the dispatcher- you’ll always dispatch to a nested tensor kernel.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 1]
```diff
@@ -560,6 +560,27 @@ Tensor NestedTensor_elementwise_Tensor(
     const Tensor& other,
     const std::string& op_name,
     Func f) {
+  // NOTE [dispatch when mixing nested and regular tensors]
+  // as of 6.24.2022, the dispatcher behaviour is:
+  // if self.is_nested() or other.is_nested() == True: dispatch to this function
+  // idk if this is the expected behaviour
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:566]: Yes, that's expected.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 2]
```diff
@@ -560,6 +560,27 @@ Tensor NestedTensor_elementwise_Tensor(
...
+  // self is a scalar
+  if (!self.is_nested() && self.numel() == 1) {
+    auto other_impl = get_nested_tensor_impl(other);
+    return wrap_buffer(
+      f(self, other_impl->get_buffer()),
+      other_impl->get_nested_size_tensor().clone()
+    );
+  }
+  // other is a scalar
+  if (!other.is_nested() && other.numel() == 1) {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:577]: You can have numel == 1 and it's not a scalar.

```
>>> torch.tensor(4.).numel()
1
>>> torch.tensor([4.]).numel()
1
```

[Scalar represents a 0-dimensional tensor which contains a single element.](https://github.com/pytorch/pytorch/blob/master/c10/core/Scalar.h#L25)

Having said that, you might not actually need to even land support for `mul.Scalar` and `mul_.Scalar` to begin with.

[`__rmul__`, `__mul__`, `__imul__` all are bound to `THPVariable_mul` and `THPVariable_mul_`](https://github.com/pytorch/pytorch/blob/f7ee061638aa2011191caeff4438fa8aff5bfec3/tools/autograd/templates/python_variable_methods.cpp#L1256-L1258), which is code generated to contain

```
  static PyObject * THPVariable_mul(PyObject* self_, PyObject* args, PyObject* kwargs)
  {
    HANDLE_TH_ERRORS
    static PythonArgParser parser({
      "mul(Tensor input, Tensor other, *, Tensor out=None)",
    }, /*traceable=*/true);

    ParsedArgs<3> parsed_args;
    auto _r = parser.parse(nullptr, args, kwargs, parsed_args);
    if(_r.has_torch_function()) {
      return handle_torch_function(_r, nullptr, args, kwargs, THPVariableFunctionsModule, "torch");
    }
    if (_r.isNone(2)) {
      // aten::mul.Tensor(Tensor self, Tensor other) -> Tensor

      auto dispatch_mul = [](const at::Tensor & self, const at::Tensor & other) -> at::Tensor {
        pybind11::gil_scoped_release no_gil;
        return self.mul(other);
      };
      return wrap(dispatch_mul(_r.tensor(0), _r.tensor(1)));
    } else {
      // aten::mul.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)

      auto dispatch_mul_out = [](at::Tensor out, const at::Tensor & self, const at::Tensor & other) -> at::Tensor {
        pybind11::gil_scoped_release no_gil;
        return at::mul_out(out, self, other);
      };
      return wrap(dispatch_mul_out(_r.tensor(2), _r.tensor(0), _r.tensor(1)));
    }
    Py_RETURN_NONE;
    END_HANDLE_TH_ERRORS
  }
```

and

```
  static PyObject * THPVariable_mul_(PyObject* self_, PyObject* args, PyObject* kwargs)
  {
    HANDLE_TH_ERRORS
    const Tensor& self = THPVariable_Unpack(self_);
    static PythonArgParser parser({
      "mul_(Tensor other)",
    }, /*traceable=*/true);

    ParsedArgs<1> parsed_args;
    auto _r = parser.parse(self_, args, kwargs, parsed_args);
    if(_r.has_torch_function()) {
      return handle_torch_function(_r, self_, args, kwargs, THPVariableClass, "torch.Tensor");
    }
    // aten::mul_.Tensor(Tensor(a!) self, Tensor other) -> Tensor(a!)

    auto dispatch_mul_ = [](const at::Tensor & self, const at::Tensor & other) -> at::Tensor {
      pybind11::gil_scoped_release no_gil;
      return self.mul_(other);
    };
    return wrap(dispatch_mul_(self, _r.tensor(0)));
    Py_RETURN_NONE;
    END_HANDLE_TH_ERRORS
  }
```

Based on that it should be enough to add a case to `NestedTensor_mul_Tensor` and `NestedTensor_mul__Tensor` that adds support for 0-dim Tensors with 1 element.

**Comment by cpuhrsch** [comment]
I'm very much in favor of Option 1. Please see my inline comment for a tour of our Python frontend bindings. I'd have also expected mul.Scalar to be used, but it seems like it's not.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp (hunk 2 above)
Comment (YifanShenSZ) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp]: scalar support is now for `dim() == 0 && numel() == 1` only

About mul.scalar, I was confused why it never got invoked. I and Joel saw a comment in `native_functions.yaml` for many .scalar functions saying that "# For C++ only, until we have conversion from C++ numbers to Tensor".

Since now we have confirmed the `THPVariable` code, I have removed mul.scalar support

**Review by cpuhrsch — APPROVED** [review]
